packages/datadog-plugin-aws-sdk/src/helpers.js

```diff
@@ -0,0 +1,172 @@
+'use strict'
+
+const helpers = {
+  finish (span, err) {
+    if (err) {
+      span.setTag('error', err)
+      this.addRequestIdTag(span, err)
+    }
+
+    span.finish()
+  },
+
+  addResponseTags (span, response, serviceName, config) {
+    if (!span) return
+
+    if (response.request) {
+      this.addServicesTags(span, response, serviceName)
+    }
+
+    this.addRequestIdTag(span, response)
+    config.hooks.request(span, response)
+  },
+
+  addServicesTags (span, response, serviceName) {
+    if (!span) return
+
+    const tags = {}
+    const params = response.request.params
+    const operation = response.request.operation
+
+    if (operation && params) {
+      switch (serviceName) {
+        case 'dynamodb':
+          this.addDynamoDbTags(params, operation, tags)
+          break
+
+        case 'kinesis':
+          this.addKinesisTags(params, operation, tags)
+          break
+
+        case 's3':
+          this.addS3Tags(params, operation, tags)
+          break
+
+        case 'sqs':
+          this.addSqsTags(params, operation, tags)
+          break
+
+        case 'sns':
+          this.addSnsTags(params, operation, response, tags)
+          break
+
+        case 'cloudwatchlogs':
+          this.addCwLogsTags(params, operation, tags)
+          break
+      }
+    }
+
+    this.addDefaultTags(operation, tags)
+    span.addTags(tags)
+  },
+
+  addDefaultTags (operation, tags) {
+    // defaults
+    if (tags['resource.name']) return
+
+    Object.assign(tags, {
+      'resource.name': operation || 'Amazon'
+    })
+  },
+
+  addDynamoDbTags (params, operation, tags) {
```

Comment:
I think one way to organize this better, especially as we continue to improve this plugin, would be to split each service in its own file with a common interface, and then this can be made a lot more dynamic.

For example:

```js
const services = {
  dynamo: require('./services/dynamo') 
}

addServicesTags (span, response, serviceName) {
  const service = services[serviceName]
  if (operation && params && service) {
    service.addTags(params, operation, tags)
  }
}
```

This could even be improved to only lazy-load helpers for services that are needed. It will also make it a lot more readable when we add things like context propagation and so on.

Reply:
yea, agreed, makes sense. I've refactored the approach here to use classes in separate files with some base class behavior. I think it's much cleaner and just generally more extensible. I've held off on the lazy loading bit for now, per our discussion, because I'd rather kick the can on worrying about any perf impact of doing synchronous file loading (though it's almost certainly trivial)

Essentially instead of doing all these seperate methods for adding different tags and that big case statemment, 

 - theres a base class which generates default tags and assigns any additional tags generated by `_addServiceTags`, which is a no-op that returns an empty object on the base class, so basically

```
    const tags = Object.assign({<default tags>}, this._addServiceTags())
```


 - service classes which define _addServiceTags with their own service specific tagging logic
- a `getService` helper function which looks up and creates an instance of the service specific class based on the service name, and, in the case it doesnt exist, uses the base class 
- Then the only actual code in the helper is

```
  const service = getService(serviceName)
  service.addTags(span, params, operation, response)
```

Definitely think its much simpler and extensible going forward. there's one weird edge case where error's will have the requestId attached, but i've just hardcoded that additional check in the error handling code. anyway lmk what you think